### DISCUSSION TITLE
Enable CORS on Vercel

### DISCUSSION THREAD

**Discussion by devpotatoes** [vercel/community#5205] [Help]
### Summary

Hi, i am building a little API tha enable me to receive a POST request and simple update my DB nothing else. But when i try it locally everything fine, and when i try it when my API is on the vercel server, error with preflight request and i don't know how to fix it...
Hope someone could help me with this problem, have good day :)

In my app.js

app.use((req, res, next) => {
    res.setHeader('Access-Control-Allow-Origin', 'https://psw-website.vercel.app');
    res.setHeader("Access-Control-Allow-Credentials", "false");
    res.setHeader('Access-Control-Allow-Headers', 'X-CSRF-Token,Origin, X-Requested-With, Content, Accept, Accept-Version, Content-Length, Content-MD5, Content-Type, Date, X-Api-Version, Authorization');
    res.setHeader('Access-Control-Allow-Methods', 'GET, POST, PUT, DELETE, PATCH, OPTIONS');
    next();
});

Error:

Access to fetch at 'https://psw-api.vercel.app/newDownload' from origin 'https://psw-website.vercel.app/' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: It does not have HTTP ok status. Ensure preflight responses are valid
[vercel.json](https://github.com/vercel/community/files/13693380/vercel.json)

### Steps to Reproduce

const feedBackInformations = {
            stars: emojiRatingIndex + 1
        };
        await fetch(`${apiURL}/newFeedback`, {
            method: "POST",
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify(feedBackInformations)
        });

**Comment by Rojojun** [comment]
I think you can solve this probleam in `next.config.js` . You should add rewrites function in the configuration source! Then CORS might be solved

**Reply by devpotatoes** [reply]
But i am not using next.js only express.

**Reply by Rojojun** [reply]
Oh! I'm sorry... to misunderstand in this case... I don't know! I'm so sorry!

**Comment by netanelben** [comment]
in express you can use the cors package:

```js
var cors = require('cors')
app.use(cors());
```

https://www.npmjs.com/package/cors

**Reply by devpotatoes** [reply]
Yeah i already tested that but i get this error
Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.

app.use(cors());
app.use((req, res, next) => {
    res.header('Access-Control-Allow-Origin', 'https://psw-website.vercel.app');
    res.header('Access-Control-Allow-Methods', "GET,OPTIONS,PATCH,DELETE,POST,PUT");
    res.header('Access-Control-Allow-Headers', "X-CSRF-Token, X-Requested-With, Accept, Accept-Version, Content-Length, Content-MD5, Content-Type, Date, X-Api-Version");
    res.header('Access-Control-Allow-Credentials', 'true');
    if (req.method === 'OPTIONS') {
        res.status(200).end();
        return;
    };
    next();
});

My vercel.json
{
    "version": 2,
    "builds": [
        {
            "src": "app.js",
            "use": "@vercel/node"
        }
    ],
    "routes": [
        {
            "src": "/(.*)",
            "dest": "app.js",
            "headers": [
                { "key": "Access-Control-Allow-Credentials", "value": "true" },
                { "key": "Access-Control-Allow-Origin", "value": "*" },
                { "key": "Access-Control-Allow-Methods", "value": "GET,OPTIONS,PATCH,DELETE,POST,PUT" },
                { "key": "Access-Control-Allow-Headers", "value": "X-CSRF-Token, X-Requested-With, Accept, Accept-Version, Content-Length, Content-MD5, Content-Type, Date, X-Api-Version" }
            ]
        }
    ]
}

**Reply by netanelben** [reply]
Which express.js version you have there ?

This would be a proper vercel.json config you should try out, 
I'm using this on a Fastify service deployed to vercel function, and it works well:

Ref: https://vercel.com/docs/projects/project-configuration#headers

```
{
  "outputDirectory": "build",
  "rewrites": [
    {
      "source": "/(.*)",
      "destination": "/api/serverless.ts"
    }
  ],
  "headers": [
    {
      "source": "/(.*)",
      "headers": [
        { "key": "Access-Control-Allow-Credentials", "value": "true" },
        { "key": "Access-Control-Allow-Origin", "value": "*" },
        {
          "key": "Access-Control-Allow-Methods",
          "value": "GET,OPTIONS,PATCH,DELETE,POST,PUT"
        },
        {
          "key": "Access-Control-Allow-Headers",
          "value": "X-CSRF-Token, X-Requested-With, Accept, Accept-Version, Content-Length, Content-MD5, Content-Type, Date, X-Api-Version"
        }
      ]
    }
  ]
}
```

**Reply by devpotatoes** [reply]
4.18.2

Now i get an other error: 405 Method Not Allowed
But i think my server is well configurated so idk why it is saying that...

{
    "outputDirectory": "build",
    "rewrites": [
        {
            "source": "/(.*)",
            "destination": "/app.js"
        }
    ],
    "headers": [
        {
            "source": "/(.*)",
            "headers": [
                {
                    "key": "Access-Control-Allow-Credentials",
                    "value": "true"
                },
                {
                    "key": "Access-Control-Allow-Origin",
                    "value": "*"
                },
                {
                    "key": "Access-Control-Allow-Methods",
                    "value": "GET,OPTIONS,PATCH,DELETE,POST,PUT"
                },
                {
                    "key": "Access-Control-Allow-Headers",
                    "value": "X-CSRF-Token, X-Requested-With, Accept, Accept-Version, Content-Length, Content-MD5, Content-Type, Date, X-Api-Version"
                }
            ]
        }
    ]
}

and here is my request:

await fetch(`${apiURL}/newDownload`, {
                method: "POST",
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify(downloadInformations)
            });

**Comment by whalyf** [comment]
First CORS-ORIGIN error, then PREFLIGHT error, i've solved that adding rewrites and many headers allowed to my ``vercel.json`` file.
My api is just a simple serveless function in golang. You guys should try that, worked for me.
```
 {
  "version": 2,
  "build": {
    "env": {
      "GO_BUILD_FLAGS": "-ldflags '-s -w'"
    }
  },
  "rewrites": [
    {
      "source": "/api/(.*)",
      "destination": "/api/main.go"
    }
  ],
  "headers": [
    {
      "source": "/api/(.*)",
      "headers": [
        { "key": "Access-Control-Allow-Credentials", "value": "true" },
        { "key": "Access-Control-Allow-Origin", "value": "*" },
        { "key": "Access-Control-Allow-Methods", "value": "GET,OPTIONS,PATCH,DELETE,POST,PUT" },
        { "key": "Access-Control-Allow-Headers", "value": "X-CSRF-Token, X-Requested-With, Accept, Accept-Version, Content-Length, Content-MD5, Content-Type, Date, X-Api-Version" }
      ]
    }
  ]
}
```
